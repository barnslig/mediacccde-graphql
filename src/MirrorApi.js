const { RESTDataSource } = require("apollo-datasource-rest");

const camelizeObj = require("./camelizeObj");

class MirrorApi extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://cdn-api.media.ccc.de/";
  }

  async getMirrors() {
    const data = await this.get("/");
    return data.MirrorList.map(m => camelizeObj(m));
  }

  async getMirror(id) {
    const mirrors = await this.getMirrors();
    return mirrors.find(m => m.id === id);
  }
}

module.exports = MirrorApi;
