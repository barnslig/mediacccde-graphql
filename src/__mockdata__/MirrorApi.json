{
  "Status": { "status": "ok" },
  "MirrorList": [
    {
      "ID": "berlin-ak",
      "HttpURL": "https://berlin-ak.ftp.media.ccc.de/",
      "Latitude": 51,
      "Longitude": 9,
      "ContinentCode": "EU",
      "CountryCodes": "DE",
      "LastSync": 1548011990,
      "Asnum": 50472,
      "SponsorURL": "",
      "SponsorLogoURL": "",
      "SponsorName": "",
      "Up": 0,
      "Enabled": true,
      "FileCount": 45255,
      "MonthDownloads": 2206722,
      "MonthBytes": 866022092309112
    },
    {
      "ID": "netcologne",
      "HttpURL": "https://mirror.netcologne.de/CCC/",
      "Latitude": 50.9333,
      "Longitude": 6.95,
      "ContinentCode": "EU",
      "CountryCodes": "DE",
      "LastSync": 1548012583,
      "Asnum": 8422,
      "SponsorURL": "http://mirror.netcologne.de/munin/netcologne.de/mirror.netcologne.de/",
      "SponsorLogoURL": "",
      "SponsorName": "Netcologne",
      "Up": 1,
      "Enabled": true,
      "FileCount": 44393,
      "MonthDownloads": 2181163,
      "MonthBytes": 835205574713437
    },
    {
      "ID": "selfnet",
      "HttpURL": "https://mirror-1.server.selfnet.de/CCC/",
      "Latitude": 48.7667,
      "Longitude": 9.1833,
      "ContinentCode": "EU",
      "CountryCodes": "DE",
      "LastSync": 1548012181,
      "Asnum": 553,
      "SponsorURL": "",
      "SponsorLogoURL": "",
      "SponsorName": "",
      "Up": 1,
      "Enabled": true,
      "FileCount": 44392,
      "MonthDownloads": 1946338,
      "MonthBytes": 755608497865107
    },
    {
      "ID": "fau",
      "HttpURL": "https://ftp.fau.de/cdn.media.ccc.de/",
      "Latitude": 49.5888,
      "Longitude": 11.0098,
      "ContinentCode": "EU",
      "CountryCodes": "DE",
      "LastSync": 1548013187,
      "Asnum": 680,
      "SponsorURL": "http://ftp.fau.de/cgi-bin/show-ftp-stats.cgi?statstype=2\u0026what=bytes\u0026mirrorname=cdn.media.ccc.de\u0026timespan=-1\u0026graphsize=large\u0026submit=Go%21",
      "SponsorLogoURL": "",
      "SponsorName": "Friedrich-Alexander-Universität Erlangen-Nürnberg",
      "Up": 1,
      "Enabled": true,
      "FileCount": 44392,
      "MonthDownloads": 1352635,
      "MonthBytes": 528296142132612
    },
    {
      "ID": "dotsrc",
      "HttpURL": "https://mirrors.dotsrc.org/cdn.media.ccc.de/",
      "Latitude": 56,
      "Longitude": 10,
      "ContinentCode": "EU",
      "CountryCodes": "DK",
      "LastSync": 1548012778,
      "Asnum": 1835,
      "SponsorURL": "",
      "SponsorLogoURL": "",
      "SponsorName": "",
      "Up": 1,
      "Enabled": true,
      "FileCount": 44392,
      "MonthDownloads": 1315870,
      "MonthBytes": 507648400389952
    },
    {
      "ID": "koeln",
      "HttpURL": "https://koeln.ftp.media.ccc.de/",
      "Latitude": 50.7374,
      "Longitude": 7.0982,
      "ContinentCode": "EU",
      "CountryCodes": "DE",
      "LastSync": 1548012979,
      "Asnum": 680,
      "SponsorURL": "https://koeln.media.ccc.de/mrtg/",
      "SponsorLogoURL": "",
      "SponsorName": "Studierendenwerk Bonn",
      "Up": 1,
      "Enabled": true,
      "FileCount": 44392,
      "MonthDownloads": 1271733,
      "MonthBytes": 495251404076822
    },
    {
      "ID": "halifax",
      "HttpURL": "https://ftp.halifax.rwth-aachen.de/ccc/",
      "Latitude": 50.7708,
      "Longitude": 6.1053,
      "ContinentCode": "EU",
      "CountryCodes": "DE",
      "LastSync": 1548012314,
      "Asnum": 47610,
      "SponsorURL": "",
      "SponsorLogoURL": "",
      "SponsorName": "",
      "Up": 1,
      "Enabled": true,
      "FileCount": 26876,
      "MonthDownloads": 995866,
      "MonthBytes": 410755417009154
    },
    {
      "ID": "oneEU",
      "HttpURL": "https://mirror.eu.oneandone.net/projects/media.ccc.de/",
      "Latitude": 51.2993,
      "Longitude": 9.491,
      "ContinentCode": "EU",
      "CountryCodes": "DE",
      "LastSync": 1548012954,
      "Asnum": 8560,
      "SponsorURL": "http://mirror.eu.oneandone.net/info.html",
      "SponsorLogoURL": "",
      "SponsorName": "1\u00261 Internet",
      "Up": 1,
      "Enabled": true,
      "FileCount": 44392,
      "MonthDownloads": 786793,
      "MonthBytes": 289558411450460
    },
    {
      "ID": "oneUS",
      "HttpURL": "https://mirror.us.oneandone.net/projects/media.ccc.de/",
      "Latitude": 37.751,
      "Longitude": -97.822,
      "ContinentCode": "NA",
      "CountryCodes": "US",
      "LastSync": 1548011739,
      "Asnum": 8560,
      "SponsorURL": "http://mirror.us.oneandone.net/info.html",
      "SponsorLogoURL": "",
      "SponsorName": "1\u00261 Internet",
      "Up": 1,
      "Enabled": true,
      "FileCount": 44392,
      "MonthDownloads": 526876,
      "MonthBytes": 194016299289951
    },
    {
      "ID": "darmstadt",
      "HttpURL": "https://ftp.darmstadt.ccc.de/",
      "Latitude": 49.8728,
      "Longitude": 8.6512,
      "ContinentCode": "EU",
      "CountryCodes": "DE",
      "LastSync": 1548011857,
      "Asnum": 8365,
      "SponsorURL": "",
      "SponsorLogoURL": "",
      "SponsorName": "",
      "Up": 1,
      "Enabled": true,
      "FileCount": 44392,
      "MonthDownloads": 61416,
      "MonthBytes": 24865146877075
    }
  ]
}
