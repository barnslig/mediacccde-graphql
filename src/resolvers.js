const { GraphQLDate, GraphQLDateTime } = require("graphql-iso-date");

const MediaApi = require("./MediaApi");

/**
 * Generates a sort function
 *
 * The resulting function can be used to sort an Array of Objects by a key
 * specified through the order param.

 * @param {string} order - Order string formatted as [key]_(ASC|DESC)
 * @returns {function}
 */
const generateSortFn = order =>
  (() => {
    if (order === "promoted") {
      // special case: EventOrder contains promoted
      return a => (a.promoted ? -1 : 1);
    }

    const [key, direction] = order.split(/_(ASC|DESC)$/);

    if (direction === "ASC") {
      return (a, b) => (Date.parse(a[key]) < Date.parse(b[key]) ? 1 : -1);
    }
    if (direction === "DESC") {
      return (a, b) => (Date.parse(a[key]) > Date.parse(b[key]) ? 1 : -1);
    }

    return 0;
  })();

module.exports = {
  Date: GraphQLDate,

  DateTime: GraphQLDateTime,

  Recording: {
    id: _source => _source.url.split("/").slice(-1)[0],

    event: async (_source, _params, { dataSources }) =>
      dataSources.MediaApi.getEvent(_source.event_url.split("/").slice(-1)[0]),

    conference: async (_source, _params, { dataSources }) =>
      dataSources.MediaApi.getConference(
        _source.conference_url.split("/").slice(-1)[0]
      )
  },

  Event: {
    conference: async (_source, _params, { dataSources }) =>
      dataSources.MediaApi.getConference(
        _source.conferenceUrl.split("/").slice(-1)[0]
      ),

    recordings: async (_source, _params, { dataSources }) => {
      let { recordings } = _source;
      if (!recordings) {
        ({ recordings } = await dataSources.MediaApi.getEvent(_source.guid));
      }

      return recordings.map(r => MediaApi.camelize(r));
    },

    relatedEvents: async (_source, { offset, limit }, { dataSources }) =>
      Promise.all(
        _source.related
          .slice(offset, offset + limit)
          .map(related => dataSources.MediaApi.getEvent(related.event_guid))
      )
  },

  Conference: {
    id: _source => _source.url.split("/").slice(-1)[0],

    events: async (_source, { order, offset, limit }, { dataSources }) => {
      let { events } = _source;
      if (!events) {
        ({ events } = await dataSources.MediaApi.getConference(_source.id));
      }

      return events
        .map(e => MediaApi.camelize(e))
        .sort(generateSortFn(order))
        .slice(offset, offset + limit);
    }
  },

  Query: {
    mirrors: async (_source, _params, { dataSources }) =>
      dataSources.MirrorApi.getMirrors(),

    mirror: async (_source, { id }, { dataSources }) =>
      dataSources.MirrorApi.getMirror(id),

    news: async (_source, { order }, { dataSources }) => {
      const data = await dataSources.NewsApi.getNews();
      return data.sort(generateSortFn(order));
    },

    conferences: async (_source, { order, offset, limit }, { dataSources }) => {
      const data = await dataSources.MediaApi.getConferences();
      return data.sort(generateSortFn(order)).slice(offset, offset + limit);
    },

    conference: async (_source, { id }, { dataSources }) =>
      dataSources.MediaApi.getConference(id),

    events: async (_source, { order, offset, limit }, { dataSources }) => {
      const data = await dataSources.MediaApi.getEvents();
      return data.sort(generateSortFn(order)).slice(offset, offset + limit);
    },

    event: async (_source, { id }, { dataSources }) =>
      dataSources.MediaApi.getEvent(id),

    eventsSearch: async (
      _source,
      { query, offset, limit },
      { dataSources }
    ) => {
      const data = await dataSources.MediaApi.searchEvents(query);
      return data.slice(offset, offset + limit);
    },

    recording: async (_source, { id }, { dataSources }) =>
      dataSources.MediaApi.getRecording(id)
  }
};
