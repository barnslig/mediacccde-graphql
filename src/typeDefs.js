const { gql } = require("apollo-server");

module.exports = gql`
  scalar Date

  scalar DateTime

  """
  This entity is a single audio/video recording from an Event/Lecture
  """
  type Recording {
    id: ID!
    filename: String
    folder: String
    height: Int
    highQuality: Boolean
    language: String
    length: Int
    mimeType: String
    recordingUrl: String
    size: Int
    state: String
    updatedAt: DateTime
    width: Int

    event: Event
    conference: Conference
  }

  enum EventOrder {
    date_ASC
    date_DESC
    duration_ASC
    duration_DESC
    promoted
    releaseDate_ASC
    releaseDate_DESC
    viewCount_ASC
    viewCount_DESC
  }

  """
  This entity is an Event/Lecture that took place during a Conference/Lecture series and consists of multiple Recordings
  """
  type Event {
    guid: ID!
    date: DateTime
    description: String
    duration: Int
    frontendLink: String
    link: String
    originalLanguage: String
    persons: [String]
    posterUrl: String
    promoted: String
    releaseDate: Date
    slug: String
    subtitle: String
    tags: [String]
    thumbnailsUrl: String
    thumbUrl: String
    timelineUrl: String
    title: String
    updatedAt: DateTime
    viewCount: Int

    conference: Conference
    recordings: [Recording]
    relatedEvents(offset: Int = 0, limit: Int = 3): [Event]
  }

  enum ConferenceOrder {
    eventLastReleasedAt_ASC
    eventLastReleasedAt_DESC
    updatedAt_ASC
    updatedAt_DESC
  }

  """
  This entity groups multiple Events/Lectures together, e.g. a Conference or Lecture Series
  """
  type Conference {
    id: ID!
    acronym: String!
    aspectRatio: String
    eventLastReleasedAt: Date
    imagesUrl: String
    logoUrl: String
    recordingsUrl: String
    scheduleUrl: String
    slug: String
    title: String
    updatedAt: DateTime
    webgenLocation: String

    events(
      order: EventOrder = date_ASC
      offset: Int = 0
      limit: Int = 10
    ): [Event]
  }

  type Query {
    conferences(
      order: ConferenceOrder = eventLastReleasedAt_ASC
      offset: Int = 0
      limit: Int = 10
    ): [Conference]

    conference(id: ID!): Conference

    events(
      order: EventOrder = date_ASC
      offset: Int = 0
      limit: Int = 10
    ): [Event]

    event(id: ID!): Event

    eventsSearch(query: String!, offset: Int = 0, limit: Int = 10): [Event]

    recording(id: ID!): Recording
  }
`;
