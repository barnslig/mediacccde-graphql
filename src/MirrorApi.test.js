import { ApolloServer, gql } from "apollo-server";

import MirrorApi from "./MirrorApi";
import MirrorApiMockData from "./__mockdata__/MirrorApi.json";

import resolvers from "./resolvers";
import typeDefs from "./schema.graphql";

const mirrorApi = new MirrorApi();
mirrorApi.get = jest.fn((path) => {
  if (path !== "/") {
    throw new Error(`Unexpected request path: ${path}`);
  }

  return MirrorApiMockData;
});

const server = new ApolloServer({
  typeDefs,
  resolvers,
  dataSources: () => ({
    MirrorApi: mirrorApi,
  }),
});

test("getMirrors", async () => {
  const mirrorsQuery = gql`
    query mirrors($offset: Int, $limit: Int) {
      mirrors(offset: $offset, limit: $limit) {
        nodes {
          id
          asnum
          continentCode
          countryCodes
          enabled
          fileCount
          httpUrl
          lastSync
          latitude
          longitude
          monthBytes
          monthDownloads
          sponsorLogoUrl
          sponsorName
          sponsorUrl
          up
        }

        pageInfo {
          hasNextPage
          hasPreviousPage
        }

        totalCount
      }
    }
  `;

  const res = await server.executeOperation({ query: mirrorsQuery });
  expect(res).toMatchSnapshot();
});

test("getMirror", async () => {
  const mirrorQuery = gql`
    fragment TestMirror on Mirror {
      id
      asnum
      continentCode
      countryCodes
      enabled
      fileCount
      httpUrl
      lastSync
      latitude
      longitude
      monthBytes
      monthDownloads
      sponsorLogoUrl
      sponsorName
      sponsorUrl
      up
    }

    query mirror($id: ID!, $nodeId: ID!) {
      mirror(id: $id) {
        ...TestMirror
      }
      node(id: $nodeId) {
        ...TestMirror
      }
    }
  `;

  const res = await server.executeOperation({
    query: mirrorQuery,
    variables: { id: "berlin-ak", nodeId: "mirror-berlin-ak" },
  });
  expect(res.data.mirror).toEqual(res.data.node);
  expect(res).toMatchSnapshot();
});
