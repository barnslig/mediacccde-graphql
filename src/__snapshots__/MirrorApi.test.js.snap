// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMirror 1`] = `
Object {
  "data": Object {
    "mirror": Object {
      "asnum": 50472,
      "continentCode": "EU",
      "countryCodes": "DE",
      "enabled": true,
      "fileCount": 45255,
      "httpUrl": "https://berlin-ak.ftp.media.ccc.de/",
      "id": "berlin-ak",
      "lastSync": "2019-01-20T19:19:50.000Z",
      "latitude": 51,
      "longitude": 9,
      "monthBytes": "866022092309112",
      "monthDownloads": 2206722,
      "sponsorLogoUrl": "",
      "sponsorName": "",
      "sponsorUrl": "",
      "up": false,
    },
    "node": Object {
      "asnum": 50472,
      "continentCode": "EU",
      "countryCodes": "DE",
      "enabled": true,
      "fileCount": 45255,
      "httpUrl": "https://berlin-ak.ftp.media.ccc.de/",
      "id": "berlin-ak",
      "lastSync": "2019-01-20T19:19:50.000Z",
      "latitude": 51,
      "longitude": 9,
      "monthBytes": "866022092309112",
      "monthDownloads": 2206722,
      "sponsorLogoUrl": "",
      "sponsorName": "",
      "sponsorUrl": "",
      "up": false,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`getMirrors 1`] = `
Object {
  "data": Object {
    "mirrors": Object {
      "nodes": Array [
        Object {
          "asnum": 50472,
          "continentCode": "EU",
          "countryCodes": "DE",
          "enabled": true,
          "fileCount": 45255,
          "httpUrl": "https://berlin-ak.ftp.media.ccc.de/",
          "id": "berlin-ak",
          "lastSync": "2019-01-20T19:19:50.000Z",
          "latitude": 51,
          "longitude": 9,
          "monthBytes": "866022092309112",
          "monthDownloads": 2206722,
          "sponsorLogoUrl": "",
          "sponsorName": "",
          "sponsorUrl": "",
          "up": false,
        },
        Object {
          "asnum": 8422,
          "continentCode": "EU",
          "countryCodes": "DE",
          "enabled": true,
          "fileCount": 44393,
          "httpUrl": "https://mirror.netcologne.de/CCC/",
          "id": "netcologne",
          "lastSync": "2019-01-20T19:29:43.000Z",
          "latitude": 50.9333,
          "longitude": 6.95,
          "monthBytes": "835205574713437",
          "monthDownloads": 2181163,
          "sponsorLogoUrl": "",
          "sponsorName": "Netcologne",
          "sponsorUrl": "http://mirror.netcologne.de/munin/netcologne.de/mirror.netcologne.de/",
          "up": true,
        },
        Object {
          "asnum": 553,
          "continentCode": "EU",
          "countryCodes": "DE",
          "enabled": true,
          "fileCount": 44392,
          "httpUrl": "https://mirror-1.server.selfnet.de/CCC/",
          "id": "selfnet",
          "lastSync": "2019-01-20T19:23:01.000Z",
          "latitude": 48.7667,
          "longitude": 9.1833,
          "monthBytes": "755608497865107",
          "monthDownloads": 1946338,
          "sponsorLogoUrl": "",
          "sponsorName": "",
          "sponsorUrl": "",
          "up": true,
        },
        Object {
          "asnum": 680,
          "continentCode": "EU",
          "countryCodes": "DE",
          "enabled": true,
          "fileCount": 44392,
          "httpUrl": "https://ftp.fau.de/cdn.media.ccc.de/",
          "id": "fau",
          "lastSync": "2019-01-20T19:39:47.000Z",
          "latitude": 49.5888,
          "longitude": 11.0098,
          "monthBytes": "528296142132612",
          "monthDownloads": 1352635,
          "sponsorLogoUrl": "",
          "sponsorName": "Friedrich-Alexander-Universität Erlangen-Nürnberg",
          "sponsorUrl": "http://ftp.fau.de/cgi-bin/show-ftp-stats.cgi?statstype=2&what=bytes&mirrorname=cdn.media.ccc.de&timespan=-1&graphsize=large&submit=Go%21",
          "up": true,
        },
        Object {
          "asnum": 1835,
          "continentCode": "EU",
          "countryCodes": "DK",
          "enabled": true,
          "fileCount": 44392,
          "httpUrl": "https://mirrors.dotsrc.org/cdn.media.ccc.de/",
          "id": "dotsrc",
          "lastSync": "2019-01-20T19:32:58.000Z",
          "latitude": 56,
          "longitude": 10,
          "monthBytes": "507648400389952",
          "monthDownloads": 1315870,
          "sponsorLogoUrl": "",
          "sponsorName": "",
          "sponsorUrl": "",
          "up": true,
        },
        Object {
          "asnum": 680,
          "continentCode": "EU",
          "countryCodes": "DE",
          "enabled": true,
          "fileCount": 44392,
          "httpUrl": "https://koeln.ftp.media.ccc.de/",
          "id": "koeln",
          "lastSync": "2019-01-20T19:36:19.000Z",
          "latitude": 50.7374,
          "longitude": 7.0982,
          "monthBytes": "495251404076822",
          "monthDownloads": 1271733,
          "sponsorLogoUrl": "",
          "sponsorName": "Studierendenwerk Bonn",
          "sponsorUrl": "https://koeln.media.ccc.de/mrtg/",
          "up": true,
        },
        Object {
          "asnum": 47610,
          "continentCode": "EU",
          "countryCodes": "DE",
          "enabled": true,
          "fileCount": 26876,
          "httpUrl": "https://ftp.halifax.rwth-aachen.de/ccc/",
          "id": "halifax",
          "lastSync": "2019-01-20T19:25:14.000Z",
          "latitude": 50.7708,
          "longitude": 6.1053,
          "monthBytes": "410755417009154",
          "monthDownloads": 995866,
          "sponsorLogoUrl": "",
          "sponsorName": "",
          "sponsorUrl": "",
          "up": true,
        },
        Object {
          "asnum": 8560,
          "continentCode": "EU",
          "countryCodes": "DE",
          "enabled": true,
          "fileCount": 44392,
          "httpUrl": "https://mirror.eu.oneandone.net/projects/media.ccc.de/",
          "id": "oneEU",
          "lastSync": "2019-01-20T19:35:54.000Z",
          "latitude": 51.2993,
          "longitude": 9.491,
          "monthBytes": "289558411450460",
          "monthDownloads": 786793,
          "sponsorLogoUrl": "",
          "sponsorName": "1&1 Internet",
          "sponsorUrl": "http://mirror.eu.oneandone.net/info.html",
          "up": true,
        },
        Object {
          "asnum": 8560,
          "continentCode": "NA",
          "countryCodes": "US",
          "enabled": true,
          "fileCount": 44392,
          "httpUrl": "https://mirror.us.oneandone.net/projects/media.ccc.de/",
          "id": "oneUS",
          "lastSync": "2019-01-20T19:15:39.000Z",
          "latitude": 37.751,
          "longitude": -97.822,
          "monthBytes": "194016299289951",
          "monthDownloads": 526876,
          "sponsorLogoUrl": "",
          "sponsorName": "1&1 Internet",
          "sponsorUrl": "http://mirror.us.oneandone.net/info.html",
          "up": true,
        },
        Object {
          "asnum": 8365,
          "continentCode": "EU",
          "countryCodes": "DE",
          "enabled": true,
          "fileCount": 44392,
          "httpUrl": "https://ftp.darmstadt.ccc.de/",
          "id": "darmstadt",
          "lastSync": "2019-01-20T19:17:37.000Z",
          "latitude": 49.8728,
          "longitude": 8.6512,
          "monthBytes": "24865146877075",
          "monthDownloads": 61416,
          "sponsorLogoUrl": "",
          "sponsorName": "",
          "sponsorUrl": "",
          "up": true,
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
      "totalCount": 10,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
